cmd/frontend/db/repos_perm.go

```diff
@@ -174,20 +163,29 @@ func getFilteredRepoNames(ctx context.Context, currentUser *types.User, repos ma
 		if err != nil {
 			return nil, err
 		}
-		for repoToVerify := range myToVerify {
-			if repoPerms, ok := perms[repoToVerify.RepoName]; ok && repoPerms[p] {
-				accepted[repoToVerify.RepoName] = struct{}{}
+
+		for _, r := range perms {
+			if r.Perms.Include(p) {
+				accepted.Add(uint32(r.Repo.ID))
 			}
 		}
+
 		// continue checking repos that didn't belong to this authz provider
 		toverify = nextToVerify
 	}
 
 	if authzAllowByDefault {
-		for r := range toverify {
-			accepted[r.RepoName] = struct{}{}
+		for _, r := range toverify {
+			accepted.Add(uint32(r.ID))
+		}
+	}
+
+	filtered = make([]*types.Repo, 0, accepted.GetCardinality())
```

Comment:
there is another perf adjustment you can make if you document that we can mutate the `repos` slice. I believe in practice this should be fine. But you can do an in-place filtering of `repos`.

```suggestion
	filtered = repos[:0]
```

Reply:
Interesting! Let me see what we get for this.

Reply:
Wait, we can't do this because we need to iterate over repos so that the returned `filtered` repos are in the same order.